src/base/http/requesthandler.cpp

```diff
@@ -0,0 +1,61 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2021  Prince Gupta <guptaprince8832@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "requesthandler.h"
+
+#include <QDateTime>
+#include <QObject>
+
+#include "base/utils/gzip.h"
+#include "connection.h"
+#include "httputils.h"
+
+using namespace Http;
+
+void Http::RequestHandler::handleConnection(Connection *socket)
+{
+    Q_ASSERT(socket->status() == Connection::PendingStatus);
```

Comment:
It's a really bad idea to have such assertion in interface method.
If it handles Connection it shouldn't restrict the state (or other properties) of this connection to be accepted as input parameter.
As I said before it doesn't handle Connection currently. It simply handles each specific request received within the Connection, but in a poorly designed form that hides the essence of things.

If you really need "new interface IConnectionHandler which only takes a Connection do as it wishes" you should pass Connection once it is constructed and allow handler to do what it wants further.
Otherwise you need other handler interface that allows correct per-request handling.

Reply:
ok, I agree with your comments about the interface, the problem is that I don't want to make `RequestHandler` a QObject. or since now it's not a pure interface maybe I can make it a QObject.

Reply:
>maybe I can make it a QObject.

Maybe...
Can you describe how your another ConnectionHandler should interacts with Connection?

Reply:
this is based on an older version of this PR https://github.com/jagannatharjun/qBittorrent/commit/76002fc51c25f92818036fa9575ed175ae3c0c1c#diff-877ca73ad30842b71bcdadd27bac68ce563ed8aa9af32a65c667ad08b8de4d9aR155

Reply:
>this is based on an older version of this PR jagannatharjun@76002fc#diff-877ca73ad30842b71bcdadd27bac68ce563ed8aa9af32a65c667ad08b8de4d9aR155

Well, let's try to figure out what the differences are between the existing handlers (web application and tracker) and the one that is needed to support media streaming.

Existing implementation just handles each request and send response. It doesn't close connection after each response. Connection can be closed either by client side or by timeout (and maybe due to underlying network error). It's "request handling" in its essence so it can be easily implemented without direct access to connection. When request is received it is passed to the handler, who processes it and returns a response that is sent to the client (i.e. all network IO can be done outside of handler).

Another handler still "just handles each request and send response" as existing one. The differences (judging by the code above) are that:
1. It closes connection after each request;
2. It handles request asynchronously.

As for (1), I have a question, is this really required, or can it re-use the existing connection to send multiple requests?
Anyway the connection can be closed outside the handler based on some Response attribute. So the main problem is (2). Even if writing to the socket in parts, mentioned by @Chocobo1, turns out to be a bad idea, and we stop at accumulating the entire Response, and only then sending it to the client, it will still not be available immediately. Apparently, here it remains only to use the Qt meta-object system, and allow the handler to return the generated Response not directly from the method, but through the signal. You only need to establish some kind of relationship between the Request, its Connection, and the Response.

Reply:
> You only need to establish some kind of relationship between the Request, its Connection, and the Response.

once I tried to make Response::content as QIODevice but it seems to complicate the file serving. https://github.com/qbittorrent/qBittorrent/blob/master/src/base/http/types.h#L121

I originally tried to make this something like Pythons; SimpleHttpServer (maybe the name is incorrect). In Python, they combine Connection and Server class and there are virtual function doHead doGet function to process the request but that would require refactoring in other parts which I am not interested in

Reply:
While I was thinking about all this and writing the previous long comment, I came up with the idea of a compromise option.
You just add a property to the Request that contains a pointer to its Connection. When the next Request is received, you pass it to the handler, so that it will still be the RequestHandler, and then it is responsible for processing the Request and sending the response through the Connection.
I.e., you do not have to change almost anything. Only rename IConnectionHandler to IRequestHandler, RequestHandler to something like BasicRequestHandler. Also you need to obtain the Request instance outside of the handler and pass it through the `IRequestHandler::handleRequest()` method.
That's all, if I didn't miss anything important.

Reply:
> You just add a property to the Request that contains a pointer to its Connection.

IMO making it property will expose connection to too many places. so I am providing connection as an argument in IRequestHander::handleRequest.

Reply:
>making it property will expose connection to too many places.

What exactly are the places that bother you? In any case, you expose it to the outside world (even through the parameter). Do you want to hide it from the inner space, where it also comes from?

Reply:
> What exactly are the places that bother you?

`BasicRequestHandler::processReqeust` `tryCompressContent`.

> In any case, you expose it to the outside world  (even through the parameter)

with parameter it is only exposed to IrequestHandler::handleRequest and doesn't leak to outside space.

Reply:
>with parameter it is only exposed to IrequestHandler::handleRequest and doesn't leak to outside space.

...unless you allow it in `IRequestHandler::handleRequest()`.
Well, let's keep it that way for now.